builtin/update-index.c

```diff
@@ -411,6 +413,12 @@ static int add_cacheinfo(unsigned int mode, const struct object_id *oid,
 	if (!verify_path(path, mode))
 		return error("Invalid path '%s'", path);
 
+	if (S_ISSPARSEDIR(mode))
+		return error("Cannot add directory '%s' as cache entry", path);
+
+	if (!path_in_cone_modesparse_checkout(path, &the_index))
+		ensure_full_index(&the_index);
```

Comment:
Did you check if this was needed? Looking into `add_cache_entry()`, it will call `index_name_stage_pos()` which will expand the index if the path is inside a sparse directory entry.

The (rare) case that I can think of is the case of a merge conflict which expands a sparse directory entry. Then, updating that entry which already exists in the index wouldn't need expansion, but this check would cause an expansion.

If you can remove this (and similar checks) without failing a test, then I think we should.

Reply:
This ended up not being needed, but required a minor change in `read_cache.c` to work properly. I was also able to remove the other checks - I had added them to be "safe", but was able to remove them after making sure all the tests still passed. It's possible that I missed test case(s), though, so if you can think of any to add, please let me know!